tensorflow/compiler/xla/service/pattern_matcher_test.cc

```diff
@@ -1102,5 +1102,75 @@ TEST_F(PatternMatcherTest, CustomCallMatchers) {
       root, m::CustomCall("test_target", m::Parameter(1), m::Parameter(0))));
 }
 
+TEST_F(PatternMatcherTest, OptionalUnaryOp) {
+  constexpr char kModuleStr[] = R"(
+    HloModule test_module
+
+    ENTRY test {
+      p0 = f32[] parameter(0)
+      cos = cosine(p0)
+      ROOT out = f32[] abs(cos)
+    }
+  )";
+  TF_ASSERT_OK_AND_ASSIGN(auto hlo_module,
+                          ParseAndReturnVerifiedModule(kModuleStr));
+  auto* root = hlo_module->entry_computation()->root_instruction();
+
+  EXPECT_TRUE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kBitcast, HloOpcode::kCos}, m::Abs())));
+  EXPECT_TRUE(
+      Match(root, m::OptionalUnaryOp({HloOpcode::kBitcast, HloOpcode::kCos},
+                                     m::Abs(m::Cos()))));
+  EXPECT_TRUE(Match(
+      root, m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kAbs}, m::Cos())));
+  EXPECT_FALSE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kAbs}, m::Bitcast())));
+  EXPECT_FALSE(Match(
+      root,
+      m::OptionalUnaryOp({HloOpcode::kCos, HloOpcode::kBitcast}, m::Cos())));
+
+  std::string description = absl::StrCat(
+      "an HloInstruction which optionally matches a unary operand with any of "
+      "opcodes {",
+      HloOpcodeString(HloOpcode::kCos), ", ",
+      HloOpcodeString(HloOpcode::kBitcast),
+      "} before matching the given pattern.");
+  std::string explanation = absl::StrCat(
+      "HloInstruction doesn't have opcode ",
+      HloOpcodeString(HloOpcode::kBitcast),
+      "\nin out = f32[] abs(f32[] cos) HloInstruction also doesn't have any of "
```

Comment:
My intent is to make this matcher output errors that look like the other matchers' errors, to reduce effort users have to make when reading these often complex error messages.

I would have expected the `in out = blah blah` to come after describing the full matcher.  Whereas right now it comes in the middle of describing the matcher.  (i.e. after checking the pattern but before checking the list of opcodes).

If we reverse the order of this -- i.e. if we check for the optional unary op first?  Then it might just work?